docs: create authentication evaluation (#63)
thoth-tech/documentation: docs/OnTrack/Deployment/Enhanced Authentication/current-and-proposed-authentication-evaluation-5.1.md

are in place within the OnTrack architecture and compare this to the proposed authentication
solution to be developed by the Enhance Authentication team. Within this evaluation, the current
authentication setup and mechanisms will be described, and then analysed in terms of the risks and
issues encompassed within such a setup. The new authentication solution which is proposed to be
developed and implemented will then be described, analysing how this new system aims to advance the
authentication capabilities of its predecessor and how it will work to mitigate the risks and issues
of the current system which it will replace. Additionally, the issues and concerns which may develop
as a result of the implementation of the proposed authentication system will be examined, including
the regulatory and compliance considerations which must be addressed during the development of the
new system.

## The State of the Current Authentication Mechanisms

### The Current Authentication Setup

Currently, the OnTrack system relies upon the authentication provided by its client, Deakin
University. The authentication mechanism utilised by Deakin University is a Single Sign On (SSO)

> system which utilises Multi-Factor Authentication (MFA) technology called “Duo Security”. As part of

Reviewer: Suggestion: Duo Security is a company and MFA is the technology that we utilise from them.

```suggestion
system which utilises Multi-Factor Authentication (MFA) technology through _Duo Security_. As part of
```

account through responding to a “login request” sent to their allocated personal device. Once this
test is passed, access is granted to the OnTrack system. In the case where the user is already
logged in to Deakin SSO, has been accessing other Deakin services and then chooses to visit the
OnTrack site, they are automatically logged in through the SSO functionality. Hence, the current
authentication system is reliant upon third-party authentication, whose MFA capabilities have been
outsourced (to Duo Security).

### Risks and Issues

An analysis of the current authentication system presented the following issues and risks:

As OnTrack relies upon Deakin SSO services, it is assumed that both the login and logout
functionalities would be handled by this service. However, even after a user has logged out Deakin
SSO and (supposedly) all concurrent accounts accessed through the single login, a user’s OnTrack
account remains logged in and accessible for a period of days (perhaps weeks). This security flaw
means that the token used to access the OnTrack account first through the Deakin SSO technology

> continues to be stored within the user’s browser for long periods of time, allowing the OnTrack

Reviewer: Nitpick:

```suggestion
continues to be stored within the user’s browser for an extended time period, allowing the OnTrack
```

account to be repeatedly accessed even when the user is not currently signed in to Deakin SSO. For
cases where the user is on a shared computer, this is a high risk for unfiltered access into the
user’s account by other actors. Currently, the only method to properly sign out of the OnTrack
system is for a user to select their avatar icon and, from the displayed drop-down menu, to choose
“sign out” from there. This is often overlooked, especially as it would be assumed that the logout
process would be handled by Deakin SSO services. Even when the user does follow the OnTrack process
to sign out of the account, the user is redirected to a broken link where the page cannot be

> displayed – the token is cleared from the browser, however, so it still works in that regard.

Reviewer: Suggestion: slight reword, less casual

```suggestion
displayed – the token is cleared from the browser. Although this does solve the issue at hand, it is not an ideal user experience
```

A second risk is that this current method of authentication relies on third-party and outsourced

> authentication technologies. Through utilising the services provided by Deakin SSO and MFA
> facilitated by Duo Security, although these technologies may be highly efficient, secure and
> reliable, such a reliance on third-party software means that the backend workings of this software

Reviewer: Suggestion: reword sentence for clarity

```suggestion
authentication technologies. Deakin SSO and MFA is facilitated by Duo Security,
although these technologies may be highly efficient, secure and reliable, such
a reliance on third-party software means that the backend workings of this software
```

is not able to be accessed and understood by the OnTrack team. Additionally, use of third-party
software requires additional sharing, transmission, and storage of user information on systems which
are not able to be managed by the OnTrack team. In the case of a security breach, the OnTrack team
are also forced to rely upon the providers of these technologies to:

- Report that the breach happened
- Report which details and systems have been compromised
- Fix the issues which lead to the breach
- Secure the system and continue normal business operations

> Having to rely upon other vendors for these processes removes the control and level which OnTrack is
> able to gain information about the scale and nature of security breaches. This leaves the company

Reviewer: Nitpick: reword 

```suggestion
Having to rely upon other vendors for these processes removes the control and information
transparency OnTrack has regarding the scale and nature of security breaches. This leaves the company
```

“in the dark” about what has occurred and is frustrating, especially when there are capable members
within the company who would be able to respond to such events, perhaps more efficiently than these
vendors.

Hence, from these risks and issues associated with the current OnTrack authentication mechanisms, it
is clear why new technologies and methods of authentication have been proposed to be developed and
implemented.

## Proposed Authentication Enhancements

This section will detail the proposed authentication elements to be added to the OnTrack
architecture in order to improve the state of the authentication mechanisms and to alleviate some of
the risks and errors in the described current setup.

### The Proposed Authentication Mechanism

> The solution which has been proposed to be implemented by the Enhance Authentication team has

Reviewer: Nitpick: simplified wording

```suggestion
The proposed solution to be implemented by the Enhance Authentication team has
```

several elements, as follows:

> - An extension of the current OnTrack API, adding code which will facilitate user management
>   functionalities
> - In line with this API extension, the addition of a feature to allow user passwords to be changed,
>   either by the user themselves or through sending a request to the admin to authorise the password
>   change. It is important that, as part of this feature, the admin is able to manipulate the
>   password (i.e. send the request for a password change) but not allow to see the plaintext password
>   at any time during this process, both before and after the password change
> - The addition of a Devise LDAP server option which handles the authentication processes for
>   OnTrack, allowing the authentication to be performed fully “in-house” rather than outsourced to
>   other authentication mechanisms

Reviewer: Nitpick: reword to simplify. (Might need formatting if you add this suggestion)

```suggestion
- An extension of the current OnTrack API, adding functionality to facilitate user management
- Password management features, for users and admin manipulation. Users can either send
a request to the admin to authorise the password or do so themselves. It is important that,
as part of this feature, the admin is able to manipulate the password (i.e. send the request
for a password change). The password is never transferred in plaintext at any time during
this process.
- A Devise LDAP server option which handles the authentication processes for OnTrack,
allowing the authentication to be performed fully “in-house” rather than outsourced to other
authentication mechanisms
```

  included as part of the authentication processes. This is an improvement as reliance on
  third-party software or processes in any circumstance increase the difficulty in understanding how
  the systems operate and what has occurred in the instance of an error or breach. Having an
  independent authentication solution streamlines the control and medication of the services
  provided by it, allowing for access and control to be granted solely to the OnTrack team on the
  backend of the project. In this way, the LDAP and Devise features used can be customised to suit
  the specific (and changing) needs of the OnTrack system, using only features that are necessary
  and relevant to the OnTrack system and user authentication needs, rather than accepting all
  services provided by third-party authentication mechanisms
- As stated, the OnTrack team will be granted full control over the authentication systems and
  elements of the OnTrack software. This will solve the issue of the redirection to the broken
  Deakin SSO link as explored in previous sections, as the team will be able to control the flow

>   towards other pages after the successful signing out of a user.

Reviewer: Suggestion: I would consider a table? Might help with readability.
Example below: You can use the text you have but format it differently

|Element |Advancement |
|---|---|
|Password management| Forgot password|
|Devise LDAP | access control|
| | |

  user, is suggested. Additionally, sending an automatic email or text to inform the user of the
  request for a change of password is also recommended
- The developed authentication system, including how user information is stored, transmitted, and
  later deleted (which includes considerations of retention laws) must adhere to the appropriate
  laws and guidelines set out by the Australian Federal Government (including the _Privacy Act
  1988_), as well as other specifications mandated within each of the Australian states and
  territories
- Additionally, as OnTrack provides its services to international users (both in terms of
  international students studying online at Deakin and through providing its services to
  international clients), OnTrack’s authentication and information storage processes must also
  adhere to the laws enacted within the relevant international jurisdictions. For example, the data
  protection laws enacted within the GDPR must be adhered to by companies who provide services to
  citizens protected under the GDPR, regardless of where the company is situated (see

>   https://gdpr-info.eu/)

Reviewer: I am so glad to see you including the GDPR! Really getting into the Global Citizenship portion of the GLOs
